src/bottom_up_iterator.jl

```diff
@@ -0,0 +1,223 @@
+"""
+	abstract type BottomUpIterator <: ProgramIterator
+
+Enumerates programs in a bottom-up fashion. This means that it starts with the smallest programs and gradually builds up to larger programs.
+The exploration of the search space is done by making use of the priority function, which associates each program with its cost.
+
+Concrete implementations of this iterator should implement the following methods:
+- `order(iter::BottomUpIterator, grammar::ContextSensitiveGrammar)::Vector{Int64}`: Returns the order in which the rules should be enumerated.
+- `pick(iter::BottomUpIterator, grammar::ContextSensitiveGrammar, state::BottomUpState, rule::Int64)::Vector{RuleNode}`: Returns the programs that can be created by applying the given rule.
+- `priority_function(iter::BottomUpIterator, grammar::ContextSensitiveGrammar, program::RuleNode, state::BottomUpState)::Int64`: Returns the priority of the given program.
+"""
+abstract type BottomUpIterator <: ProgramIterator end
+
+Base.@doc """
+    @programiterator DepthIterator(problem::Problem{Vector{IOExample}}) <: BottomUpIterator
+
+A basic implementation of the bottom-up iterator. It will enumerate all programs in increasing order based on their depth.
+""" DepthIterator
+@programiterator DepthIterator(problem::Problem{Vector{IOExample}}) <: BottomUpIterator
+
+mutable struct BottomUpState
+    bank::Any
+    data::Any
+    hash_set::Set{UInt}
+end
+
+mutable struct DepthIteratorData
+    next_programs_iterable::Any
+    next_programs_result::Union{Nothing, Tuple{Tuple{Int, Any}, Any}}
+
+    new_programs::Vector{RuleNode}
+
+    rules::Queue{Int}
+    depth::Int
+end
+
+"""
+	priority_function(iter::DepthIterator, program::RuleNode)::Int64
+
+Returns the depth of the RuleNode that describes the given program.
+"""
+function cost_function(
+    iter::DepthIterator,
```

Comment:
If there are functions that we expect users to provide for the inner `DepthIterator`, maybe we should introduce a specific type for the inner iterator of the bottom-up search. But we also then have to make sure to "couple" the bottom-up iterator (including custom defined ones) with the inner one from the signature.